src/libraries/System.Private.CoreLib/src/System/IO/RandomAccess.Windows.cs

```diff
@@ -413,90 +417,130 @@ internal static void WriteGatherAtOffset(SafeFileHandle handle, IReadOnlyList<Re
             }
         }
 
-        private static ValueTask<long> ReadScatterAtOffsetAsync(SafeFileHandle handle, IReadOnlyList<Memory<byte>> buffers,
-            long fileOffset, CancellationToken cancellationToken)
+        // From https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-readfilescatter:
+        // "The file handle must be created with [...] the FILE_FLAG_OVERLAPPED and FILE_FLAG_NO_BUFFERING flags."
+        private static bool CanUseScatterGatherWindowsAPIs(SafeFileHandle handle)
+            => handle.IsAsync && ((handle.GetFileOptions() & SafeFileHandle.NoBuffering) != 0);
+
+        // From the same source:
+        // "Each buffer must be at least the size of a system memory page and must be aligned on a system
+        // memory page size boundary. The system reads/writes one system memory page of data into/from each buffer."
+        // This method returns true if the buffers can be used by
+        // the Windows scatter/gather API, which happens when they are:
+        // 1. aligned at page size boundaries
+        // 2. exactly one page long each (our own requirement to prevent partial reads)
+        // 3. not bigger than 2^32 - 1 in total
+        // This function is also responsible for pinning the buffers if they
+        // are suitable and they must be unpinned after the I/O operation completes.
+        // It also returns a pointer with the segments to be passed to the
+        // Windows API, and the total size of the buffers that is needed as well.
+        // The pinned MemoryHandles and the pointer to the segments must be cleaned-up
+        // with the CleanupScatterGatherBuffers method.
+        private static unsafe bool TryPrepareScatterGatherBuffers<T, THandler>(IReadOnlyList<T> buffers,
+            THandler handler, out MemoryHandle[] handlesToDispose, out IntPtr segmentsPtr, out int totalBytes)
+            where THandler: struct, IMemoryHandler<T>
         {
-            if (!handle.IsAsync)
-            {
-                return ScheduleSyncReadScatterAtOffsetAsync(handle, buffers, fileOffset, cancellationToken);
-            }
+            int pageSize = s_cachedPageSize;
+            Debug.Assert(BitOperations.IsPow2(pageSize), "Page size is not a power of two.");
+            // We take advantage of the fact that the page size is
+            // a power of two to avoid an expensive modulo operation.
+            long alignedAtPageSizeMask = pageSize - 1;
+
+            int buffersCount = buffers.Count;
+            handlesToDispose = new MemoryHandle[buffersCount];
```

Comment:
Can we hold off on allocating this until we actually need to store something into it? L472

Reply:
Why? That allocation would have been useless only if we pass an empty buffer list, and I guess we can assume the power users this API caters to will not do that.

Reply:
Because there are code paths before that which return false, e.g. if the first buffer isn't properly aligned.  Why allocate in that case?

I expect it's going to be way more likely that this API will be used by call sites that don't meet the requirements than by those that do.

Reply:
Even if the method returns false, we have to somewhere store the `MemoryHandles` so that we can later dispose them. We can do something like the following:

``` csharp
MemoryHandle handle = handler.Pin(in buffer);
long ptr = segments[i] = (long)handle.Pointer;
if ((ptr & alignedAtPageSizeMask) != 0)
{
    handle.Dispose();
    return false;
}
(handlesToDispose ??= new MemoryHandle[buffersCount])[i] = handle;
```

But it saves an array allocation only if the _first_ buffer is unacceptable. How about renting from the `ArrayPool`?

And in either case we have to cleary state these requirements on Windows, in the API's documentation.

Reply:
> But it saves an array allocation only if the first buffer is unacceptable. 

Yes,  that's my suggestion. 

> How about renting from the ArrayPool?

I doubt it's worthwhile.

Reply:
> And in either case we have to cleary state these requirements on Windows, in the API's documentation.

It won't matter what the docs say, especially now that this change makes it work regardless.